src/algorithms/digi/RandomNoise.cc

```diff
@@ -0,0 +1,562 @@
+        const double y   = rMid * std::sin(phi);
+        unique.insert(seg.cellID({x, y, 0}, {x, y, 0}, 0));
+      }
+    }
+  }
+
+  /* -------------------------------------------------------------- */
+  /* PolarGridRPhi  (uniform ΔR, Δφ)                                */
+  /* -------------------------------------------------------------- */
+  template <typename POLAR_UNI>
+  void scanPolarUniform(const POLAR_UNI& seg, const TGeoBBox* box,
+                        std::set<dd4hep::CellID>& unique) {
+    const double dR     = seg.gridSizeR();
+    const double dPhi   = seg.gridSizePhi();
+    const double offR   = seg.offsetR();
+    const double offPhi = seg.offsetPhi();
+
+    const double Rmax = std::hypot(box->GetDX(), box->GetDY());
+
+    const std::size_t nR   = std::max<std::size_t>(1, std::ceil((Rmax - offR) / dR));
+    const std::size_t nPhi = std::max<std::size_t>(1, std::round(2 * M_PI / dPhi));
+
+    for (std::size_t iR = 0; iR < nR; ++iR) {
+      const double r = offR + (iR + 0.5) * dR;
+      for (std::size_t iP = 0; iP < nPhi; ++iP) {
+        const double phi = offPhi + (iP + 0.5) * dPhi;
+        const double x   = r * std::cos(phi);
+        const double y   = r * std::sin(phi);
+        unique.insert(seg.cellID({x, y, 0}, {x, y, 0}, 0));
+      }
+    }
+  }
+
+  /* Monte-Carlo fallback */
+  void scanGeneric(const dd4hep::Segmentation& seg, const TGeoBBox* box,
+                   std::set<dd4hep::CellID>& unique) {
+    static thread_local std::mt19937_64 rng{0xFEDCBA98u};
+    std::uniform_real_distribution<> ux(-box->GetDX(), box->GetDX());
+    std::uniform_real_distribution<> uy(-box->GetDY(), box->GetDY());
+    std::uniform_real_distribution<> uz(-box->GetDZ(), box->GetDZ());
+
+    for (unsigned i = 0; i < 20000; ++i) {
+      dd4hep::Position p{ux(rng), uy(rng), uz(rng)};
+      unique.insert(seg.cellID(p, p, 0));
+      if (unique.size() > 60000)
+        break;
+    }
+  }
+
+} // anonymous namespace
+
+//====================================================================//
+//  0.1  Initialisation                                                //
+//====================================================================//
+void RandomNoise::init(const dd4hep::Detector* det) { m_dd4hepGeo = det; }
+
+//====================================================================//
+//  1.  Event processing                                               //
+//====================================================================//
+void RandomNoise::process(const Input& in, const Output& out) const {
+  const auto [in_hits] = in;
+  auto [out_hits]      = out;
+
+  //----------------------------------------------------------------
+  // 1.1  Resolve the DD4hep read-out that belongs to the collection
+  //----------------------------------------------------------------
+  std::string readoutname = m_cfg.readout_name;
+
+  dd4hep::Readout rd = m_dd4hepGeo->readout(readoutname);
```

Comment:
All of the readout and geometry scanning could be moved to `init()` so it is only calculated once.